lib/concurrent-ruby/concurrent/utility/processor_counter.rb

```diff
@@ -21,6 +22,23 @@ def physical_processor_count
         @physical_processor_count.value
       end
 
+      def usable_processor_count
```

Comment:
I have no strong opinion on the name.

> And that maybe shouldn't count hyperthreads (because they very rarely give parallelism), i.e. be based on physical_processor_count? I'm unsure, WDYT?

Unfortunately I'm afraid this ship has sailed, when working on the cloud, it's all logical cores (vCPU), and unless I'm mistaken, the cgroup cpu_quota is in logical cores too, so I think it would be incorrect to only use physical ones.

Reply:
> Unfortunately I'm afraid this ship has sailed, when working on the cloud, it's all logical cores (vCPU)

Argh, indeed that seems to be the case.
At least if we expose `cpu_quota` people can use it with either `physical_processor_count` or `processor_count` depending if they want true parallelism count or basically that x 2 if hyperthreaded.

Reply:
> At least if we expose `cpu_quota` people can use it with either `physical_processor_count` or `processor_count` depending if they want true parallelism count or basically that x 2 if hyperthreaded.

Unfortunately that won't be easy because the quota is a number of CPUs and not a ratio, so one needs to know if the quota is in logical or physical cores.

Reply:
I played a bit with it and `podman run --cpus=2 --rm -it fedora:latest` with 2 `bash -c 'while true; do true; done'  &` definitely uses 2 real cores. But I guess in general one never knows and it would in practice depend what else is running on the host.

Reply:
https://docs.sylabs.io/guides/latest/user-guide/cgroups.html#cpu-limits seem to indicate that `cpu.max` is indeed expressed in physical cores. I'm starting to have a bit of a headache here...

Reply:
> https://docs.sylabs.io/guides/latest/user-guide/cgroups.html#cpu-limits seem to indicate that `cpu.max` is indeed expressed in physical cores. I'm starting to have a bit of a headache here...

I'm not sure, it's really not clear :sweat: 

I guess what docker/podman --cpus sets is the cgroup `--cpus` given it can be a fraction? But maybe there is x2 in between?

The linked docs says:
> --cpus sets the number of CPUs, or fractional CPUs, that the container can use. The minimum is 0.01 or one hundredth of a physical CPU.
> The maximum is the number of CPU cores on your system.

So if by CPU cores they mean real core, ignoring hyper threads then indeed it's physical cores.

But, I just tried with `--cpus=8` and `--cpus=16` on my 8 core + 8 hyper threads AMD machine, and with `--cpus=8` it uses approximately 50% of all CPU (looking at htop and `Load average` which was ~8), while with `--cpus=16` it uses everything (`Load average` which was ~16).
So I think it is logic processor count in the end for the Podman (and I guess Docker too) --cpus argument.

Reply:
So I think the code is fine as-is and we should use logical, and document it.

Reply:
Oh, thanks for testing. I guess none of that documentation can be trusted -_-.